lib/src/main/kotlin/StatelessSolver.kt

```diff
@@ -1,111 +1,75 @@
-import kotlin.math.ln
 import kotlin.math.max
-import kotlin.math.sqrt
-import kotlin.random.Random
-import kotlin.text.StringBuilder
 
-class StatelessSolver<TState, TAction>(
+open class StatelessSolver<TState, TAction>(
         private val mdp: MDP<TState, TAction>,
-        private val random: Random,
-        private val iterations: Int,
         private val simulationDepthLimit: Int,
-        private val explorationConstant: Double,
+        explorationConstant: Double,
         private val rewardDiscountFactor: Double,
-        private val verbose: Boolean) {
+        verbose: Boolean)
+    : SolverBase<TAction, StatelessActionNode<TState, TAction>>(verbose, explorationConstant) {
 
-    private var root : StateActionNode<TAction>? = null
-
-    fun buildTree(): MutableList<Double> {
-        initialize()
-
-        var rewardTracker = mutableListOf<Double>()
-
-        for (i in 0..iterations) {
-            iterateStep()
-
-            println(root!!.children.toString())
-            println(root!!.children.maxByOrNull { a -> calculateUCT(a)} )
-            var ns = root!!.children.maxByOrNull { a -> calculateUCT(a)}!!.n
-            var explorationFactor = explorationConstant*sqrt(ln(i.toDouble())/ns)
-            rewardTracker.add(explorationFactor)
+    override var _root = StatelessActionNode<TState, TAction>(null, null)
 
+    override fun selectNode(node: StatelessActionNode<TState, TAction>) : StatelessActionNode<TState, TAction> {
+        // If this node is a leaf node, return it
+        if (node.getChildren().isEmpty()) {
+            return node
         }
-        return rewardTracker
-    }
-
-    fun initialize() {
-        root = createStateActionNode(null, null)
-    }
 
-    fun iterateStep() {
-        traceln("")
-        traceln("New iteration")
-        traceln("=============")
+        val currentNode = node
+        simulateActions(node)
 
-        // Selection
-        val best = selectNode(root!!)
+        // Run a simulation greedily
+        while (true) {
+            if (mdp.isTerminal(currentNode.state)) {
```

Comment:
Note to @JunTaoLuo , explain why now there is a SolverBase, in addition to Stateless/fulSolver.

Reply:
SolverBase contains shared implementations that can be reused between Stateless/fulSolvers. Keeping in line with the DRY principle.